README.md
# rp2biosensor -- Build Sensing-Enabling Metabolic Pathways from RetroPath2.0 output

rp2biosensor extracts metabolic paths linking undetectable compounds
into a detectable one using biochemical reactions. This concept is known
as "Sensing-Enabling Metabolic Pathways". See [doi: 10.1021/acssynbio.5b00225](https://doi.org/10.1021/acssynbio.5b00225)
for more.

In short, rp2bionsensor converts the output of [RetroPath2.0](https://www.myexperiment.org/workflows/4987.html) into a HTML page showing the possible metabolic paths linking the compound to be detected to the detectable ones.

Notice: only the shortest paths are kept. This mean that if there is a 1 step path, no 2 steps will be outputted.

## Install

```bash
conda install -c conda-forge rp2biosensor
```

## Run

```bash
python -m rp2biosensor /path/to/rp2/results.csv --opath /path/to/output/file.html --otype file
```

By default, all the needed dependancies are embedded into the HTML file. This includes CSS, JavaScript, and data files. Alternatively, the output could be outputted into a directory, using the `dir` output type:

```bash
python -m rp2biosensor /path/to/rp2/results.csv --opath /path/to/output --otype dir
```

The embedded help:
```bash
python -m rp2biosensor -h

usage: rp2biosensor [-h] [--opath OPATH] [--otype {dir,file}] [--ojson OJSON] [--cache-dir CACHE_DIR] rp2_results sink_file

Generate HTML outputs to explore Sensing Enabling Metabolic Pathway from RetroPath2 results.

positional arguments:
  rp2_results           RetroPath2.0 results
  sink_file             Sink file used for RetroPath2.0

options:
  -h, --help            show this help message and exit
  --opath OPATH         Output path. Default: /Users/tduigou/code/rp2biosensor/biosensor.html.
  --otype {dir,file}    Output type. This could be either (i) "dir" which means ouput files will outputted into this directory, or (ii) "file" which means that all files will be embedded into a single HTML page. Default: file
  --ojson OJSON         Output the graph as json file if the path is not None. Default: None
  --cache-dir CACHE_DIR
                        Path to the cache directory. If not specified, None is passed to rrCache.
```

## Example

```bash
python -m rp2biosensor tests/data/input/rp2-results_lactate.csv --opath ./biosensor.html
```

## For developpers

### Install
```bash
conda env create -f environment.yaml -n rp2biosensor-dev
conda develop -n rp2biosensor-dev .
```

### Test
```bash
python -m pytest -vv
```

## Authors
- Thomas Duigou ([tduigou](https://github.com/tduigou))

## Licence
rp2biosensor is released under the MIT licence. See the [LICENSE.md](LICENSE.md) file for details.

CHANGELOG.md
# Changelog

## 3.2.1 (2023-01-03)

- fix(test_main.py): fix UnicodeDecodeError with windows OS

## 3.2.0 (2023-01-03)

- feat: add cache_dir arg
- docs(README.md): update usage
- test(test_main.py): fix mismatch due to SVG tags

## 3.1.0 (2022-03-23)

- fix(RP2Objects): adapt to rxn_rebuild output structure changes

## 3.0.0 (2022-03-04)

- fix(RP2Objects)!: sink matching issue

## 2.1.0 (2022-02-24)

- feat: update the default list of ignored compounds
- feat: export graph as json

## 2.0.0 (2022-02-11)

- feat: add co-substrates and co-products
- feat: SMILES of compounds are canonized
- feat: merge similar reactions into a single node
- feat: relax inchi search for compound to ignore
- fix!: change default directionnality of reactions
- fix: handle multiple usage of a same rule

## 1.0.1 (2021-12-20)

- fix endline for windows files

## 1.0.0

- first working version

tests/test_main.py
from types import SimpleNamespace
from pathlib import Path
from filecmp import cmpfiles
from rp2biosensor.__main__ import run

RESULT_PATH = Path(__file__).resolve().parent/ 'data' / 'input' / 'rp2-results-lactate.csv'
SINK_PATH = Path(__file__).resolve().parent / 'data' / 'input' / 'sink-h2o2.csv'
OUTPUT_DIR_PATH = Path(__file__).resolve().parent / 'data' / 'output_dir'
OUTPUT_FILE_PATH = Path(__file__).resolve().parent / 'data' / 'output_file'

# Compare except with expected file except for SVGs
def filter_lines(line):
    """Filter unwanted lines
    
    This method is used when filtering lines with the filter method.
    """
    if line.lstrip().startswith('"svg": '):
        return False
    return True

def compare_files(ref_dir: Path, test_dir: Path, files_to_cmp: list):
    """Compare two set of files
    
    Files are compared 2 by 2, coming from 2 different directories
    """
    match = []
    mismatch = []
    for file in files_to_cmp:
        with open(ref_dir / file, encoding="utf-8") as ref, \
             open(test_dir / file, encoding="utf-8") as test:
            f1 = filter(filter_lines, ref)
            f2 = filter(filter_lines, test)
            if all(x == y for x,y in zip(f1, f2)):
                match.append(file)
            else:
                mismatch.append(file)
    return match, mismatch


def test_dir_output(tmpdir):
    temp_path = tmpdir / 'dir_case'  # tmpdir scope is session wised
    options = {
        "rp2_results": f"{RESULT_PATH}",
        "sink_file": f"{SINK_PATH}",
        "opath": f"{temp_path}",
        "otype": "dir",
        "ojson": None,
        "cache_dir": None
        }
    files_to_cmp = [
        'index.html',
        'network.json',
        'css/viewer.css',
        'js/viewer.js'
    ]
    args = SimpleNamespace(**options)
    run(args)

    match, mismatch = compare_files(
        ref_dir=OUTPUT_DIR_PATH,
        test_dir=temp_path,
        files_to_cmp=files_to_cmp
        )
    
    try:
        assert all(item in match for item in files_to_cmp)
    except AssertionError as e:
        print("Matched Files    : {}".format(match))
        print("Mismatched Files : {}".format(mismatch))
        raise e

def test_file_output(tmpdir):
    temp_path = tmpdir / 'file_case'  # tmpdir scope is session wised
    options = {
        "rp2_results": f"{RESULT_PATH}",
        "sink_file": f"{SINK_PATH}",
        "opath": f'{temp_path / "biosensor.html"}',
        "otype": "file",
        "ojson": None,
        "cache_dir": None
        }
    files_to_cmp = ['biosensor.html']
    args = SimpleNamespace(**options)
    run(args)

    match, mismatch = compare_files(ref_dir=OUTPUT_FILE_PATH, test_dir=temp_path, files_to_cmp=files_to_cmp)
    try:
        assert 'biosensor.html' in match
    except AssertionError as e:
        print("Matched Files    : {}".format(match))
        print("Mismatched Files : {}".format(mismatch))
        raise e

tests/test_RP2Objects.py
from pathlib import Path
import pytest

from rp2biosensor import RP2Objects


@pytest.fixture
def smiles_to_test():
    return [
        # (not yet canonized, canonized)
        ('[H]OC(=O)C([H])(O[H])C([H])([H])[H]','CC(O)C(=O)O'),
        ('[H]O[H]', 'O'),
        ('[H]OC([H])=O', 'O=CO'),
    ]


def test_canonize_smiles(smiles_to_test):
    for case in smiles_to_test:
        query, ans = case
        assert RP2Objects.canonize_smiles(query) == ans


class TestIDsHandler:
    def test_new_id(self):
        idh = RP2Objects.IDsHandler()
        assert idh.make_new_id() == 'ID_0000000001'


class TestCompound:

    smi = '[H]O[H]'

    def test_init_ids(self):
        # One should initialize the ID handler
        with pytest.raises(AttributeError):
            c = RP2Objects.Compound(smiles=self.smi)

    def test_init(self):
        # This is better
        RP2Objects.Compound.init_id_handler()
        c = RP2Objects.Compound(smiles=self.smi)
        assert c.uid == 'CMPD_0000000001'
    
    def test_compute_struct(self):
        RP2Objects.Compound.init_id_handler()
        c = RP2Objects.Compound(smiles=self.smi)
        c.compute_structures()
        assert c.smiles == 'O'

    def test_cids(self):
        RP2Objects.Compound.init_id_handler()
        c = RP2Objects.Compound()
        c.add_cid('B')
        c.add_cid('A')
        assert c.get_cids() == ['A', 'B']


class TestTransformatoin:
    # To be done
    pass


class TestRP2parser:
    # To be done
    pass


class TestRetroGraph:
    # To be done
    pass

rp2biosensor/Utils.py
from pathlib import Path
from distutils import dir_util
from tempfile import TemporaryDirectory


def __write_json(json_path, json_str):
    with open(json_path, "w", newline="\n") as ofh:
        ofh.write(json_str)


def write(args, template_dir, json_str: str):
    if args.otype == 'dir':
        # Prepare output dir
        outdir_path = Path(args.opath).resolve()
        outdir_path.mkdir(parents=True, exist_ok=True)
        # Copy template dir
        dir_util.copy_tree(str(template_dir), str(outdir_path))
        # Append network
        with open(outdir_path / 'network.json', 'w', newline='\n') as ofh:
            ofh.write(f'network = {json_str}')
    elif args.otype == 'file':
        # Prepare output dir
        outdir_path = Path(args.opath).resolve().parent
        outdir_path.mkdir(parents=True, exist_ok=True)
        outfile_path = Path(args.opath).resolve()
        # Copy template dir into a temporary folder
        with TemporaryDirectory() as temp_dir:
            tempdir_path = Path(temp_dir)
            dir_util.copy_tree(str(template_dir), str(tempdir_path))
            with open(tempdir_path / 'network.json', 'w', newline='\n') as ofh:
                ofh.write(f'network = {json_str}')
            html_str = all_in_one_file(tempdir_path)
        with open(outfile_path, 'wb') as ofh:
            ofh.write(html_str.replace(b'\r\n', b'\n'))
    else:
        raise NotImplementedError(f'Unexpected otype: {args.otype}')

    if args.ojson is not None:
        __write_json(Path(args.ojson), json_str)


def all_in_one_file(ifolder: Path) -> str:
    """Merge all needed file into a single HTML

    :param ifolder: folder containing the files to be merged
    :return html_str: string, the HTML
    """
    # find and open the index file
    html_str = open(ifolder / 'index.html', 'rb').read()
    # open and read JS files and replace them in the HTML
    js_to_replace = [
        'js/jquery-3.6.0.min.js',
        'js/cytoscape-3.19.0.min.js',
        'js/jquery-ui-1.12.1.min.js',
        'js/viewer.js'
    ]
    for js_file in js_to_replace:
        js_bytes = open(ifolder / js_file, 'rb').read()
        ori = b'src="' + js_file.encode() + b'">'
        rep = b'>' + js_bytes
        html_str = html_str.replace(ori, rep)
    # open and read style.css and replace it in the HTML
    css_to_replace = ['css/viewer.css']
    for css_file in css_to_replace:
        css_str = open(ifolder / css_file, 'rb').read()
        ori = b'<link href="' + css_file.encode() + b'" rel="stylesheet" type="text/css"/>'
        rep = b'<style type="text/css">' + css_str + b'</style>'
        html_str = html_str.replace(ori, rep)
    ### replace the network
    net_str = open(ifolder / 'network.json', 'rb').read()
    ori = b'src="' + 'network.json'.encode() + b'">'
    rep = b'>' + net_str
    html_str = html_str.replace(ori, rep)
    return html_str


rp2biosensor/__main__.py
"""Build a graph representation of the retrosynthetic network."""

from __future__ import annotations

import sys
import logging
import argparse
from pathlib import Path

from rp2biosensor.RP2Objects import RP2parser
from rp2biosensor.RP2Objects import RetroGraph
from rp2biosensor.Utils import write, __write_json

# Path to templates 
TEMPLATE_DIR = Path(__file__).resolve().parent / "templates"

TARGET_ID = "TARGET_0000000001"  # Default target ID
COFACTORS = [                    # Cofactor structures to be filtered
    "InChI=1S/O2/c1-2",  # O2
    "InChI=1S/H2O/h1H2",  # Water
    "InChI=1S/p+1",  # H+
    "InChI=1S/CO2/c2-1-3",  # CO2
    "InChI=1S/C10H16N5O13P3/c11-8-5-9(13-2-12-8)15(3-14-5)10-7(17)6(16)4(26-10)1-25-30(21,22)28-31(23,24)27-29(18,19)20/h2-4,6-7,10,16-17H,1H2,(H,21,22)(H,23,24)(H2,11,12,13)(H2,18,19,20)",  # ATP
    "InChI=1S/C21H28N7O17P3/c22-17-12-19(25-7-24-17)28(8-26-12)21-16(44-46(33,34)35)14(30)11(43-21)6-41-48(38,39)45-47(36,37)40-5-10-13(29)15(31)20(42-10)27-3-1-2-9(4-27)18(23)32/h1-4,7-8,10-11,13-16,20-21,29-31H,5-6H2,(H7-,22,23,24,25,32,33,34,35,36,37,38,39)",  # NADP(+)
    "InChI=1S/C21H30N7O17P3/c22-17-12-19(25-7-24-17)28(8-26-12)21-16(44-46(33,34)35)14(30)11(43-21)6-41-48(38,39)45-47(36,37)40-5-10-13(29)15(31)20(42-10)27-3-1-2-9(4-27)18(23)32/h1,3-4,7-8,10-11,13-16,20-21,29-31H,2,5-6H2,(H2,23,32)(H,36,37)(H,38,39)(H2,22,24,25)(H2,33,34,35)",  # NADPH
    "InChI=1S/C10H15N5O10P2/c11-8-5-9(13-2-12-8)15(3-14-5)10-7(17)6(16)4(24-10)1-23-27(21,22)25-26(18,19)20/h2-4,6-7,10,16-17H,1H2,(H,21,22)(H2,11,12,13)(H2,18,19,20)",  # ADP
    "InChI=1S/C21H27N7O14P2/c22-17-12-19(25-7-24-17)28(8-26-12)21-16(32)14(30)11(41-21)6-39-44(36,37)42-43(34,35)38-5-10-13(29)15(31)20(40-10)27-3-1-2-9(4-27)18(23)33/h1-4,7-8,10-11,13-16,20-21,29-32H,5-6H2,(H5-,22,23,24,25,33,34,35,36,37)",  # NAD+
    "InChI=1S/C21H29N7O14P2/c22-17-12-19(25-7-24-17)28(8-26-12)21-16(32)14(30)11(41-21)6-39-44(36,37)42-43(34,35)38-5-10-13(29)15(31)20(40-10)27-3-1-2-9(4-27)18(23)33/h1,3-4,7-8,10-11,13-16,20-21,29-32H,2,5-6H2,(H2,23,33)(H,34,35)(H,36,37)(H2,22,24,25)",  # NADH
    "InChI=1S/C21H36N7O16P3S/c1-21(2,16(31)19(32)24-4-3-12(29)23-5-6-48)8-41-47(38,39)44-46(36,37)40-7-11-15(43-45(33,34)35)14(30)20(42-11)28-10-27-13-17(22)25-9-26-18(13)28/h9-11,14-16,20,30-31,48H,3-8H2,1-2H3,(H,23,29)(H,24,32)(H,36,37)(H,38,39)(H2,22,25,26)(H2,33,34,35)",  # CoA
]


def build_args_parser(prog='rp2biosensor'):
    desc = "Generate HTML outputs to explore Sensing Enabling Metabolic Pathway from RetroPath2 results."
    parser = argparse.ArgumentParser(description=desc, prog=prog)
    parser.add_argument('rp2_results',
                        help='RetroPath2.0 results')
    parser.add_argument("sink_file",
                        help="Sink file used for RetroPath2.0")
    # parser.add_argument('--reverse_direction',
    #                     help='Reverse direction of reactions described in RetroPath2.0 results.',
    #                     default=False, type=lambda x: (str(x).lower() == 'true'))
    parser.add_argument('--opath',
                        help=f'Output path. Default: {Path().resolve() / "biosensor.html"}.',
                        default=Path().resolve() / 'biosensor.html')
    parser.add_argument('--otype',
                        help='Output type. This could be either (i) "dir" which means '
                             'ouput files will outputted into this directory, or (ii) '
                             '"file" which means that all files will be embedded into '
                             'a single HTML page. Default: file',
                        default='file', choices=['dir', 'file'])
    parser.add_argument("--ojson",
                        help="Output the graph as json file if the path is not None. "
                             "Default: None")
    parser.add_argument("--cache-dir",
                        help="Path to the cache directory. If not specified, "
                             "None is passed to rrCache.",
                        default=None)
    return parser


def run(args):
    # Extract and build graph
    rparser = RP2parser(args.rp2_results, args.sink_file, cache_dir=args.cache_dir)
    rgraph = RetroGraph(rparser.compounds, rparser.transformations)
    nb_paths = rgraph.keep_source_to_sink(to_skip=COFACTORS, target_id=TARGET_ID)
    rgraph.refine()
    # Write output
    json_str = rgraph.to_cytoscape_export()
    write(args, TEMPLATE_DIR, json_str)


def main():
    logging.basicConfig(
        stream=sys.stderr, level=logging.INFO,
        datefmt='%d/%m/%Y %H:%M:%S',
        format='%(asctime)s -- %(levelname)s -- %(message)s'
    )
    parser = build_args_parser()
    args = parser.parse_args()
    run(args)


if __name__ == "__main__":
    main()